### PR TITLE
feat: zero-knowledge net 2

### PR DESCRIPTION
**Pull request by jorgenhanssen**
I wanted to create another zero-knowledge net with some slight adjustments from the one made in #189.

Most notably:

**1. Latest/greatest search:** The search has been improved since the last zero-knowledge net and I expect better/deeper evaluations.

**2. Fixed node budgets:** The previous nets have been trained on a fixed depth 8 search. I expect a fixed node budget to be better since early game positions are rather drawish and depth is much more needed in endgames (pawn races, K+K+B  vs K endgames, etc)

**3. Endgame tablebases:** This should ensure better WDL qualities. I did see grail struggle and often over-estimate drawn positions. I could also find this over-confidence further up the tree, and while I don't know if this was the cause, I will run a full re-train just to be sure since this change with the existing overconfident search + 20% WDL blend is not gonna cut it (I think).

**4. Dense samples:** I want to try to train on all samples within a game instead of just the sparse anchor/decision nodes. The reason is that while skipping intermediate nodes allows for a much higher variety of games per sample, we do miss out on crucial nuances within the game, especially in endgames that have specific sequences. Also, the throughput is higher since the search is allowed to run on sequantial positions, which keeps more relevant entries in the TT and history - so with a fixed node budget the searches should be deeper/better.

**5. Decaying WDL blend:** In the other zero-net training I did a cut from 100% to 50% WDL after the seed net. This might have introduced slight bad biases into the search (?). At least that is my intuition, so this time I will start at 100% and work my way down in steps of 10%.

**6. More samples per iteration:** Yeah gonna crank up the number of samples per iteration on this net. Gonna take longer but it will hopefully help it generalize better. Also without the sparse sampling we need more samples per game, so. 

**7. Random 8-ply openings:** The last net was trained with openings from a book. Felt reasonable back then but I have a hunch this is not the best way to go about this, and that is probably why most datagen in engines uses a random opening - so i will as well.

Other than that, the run will be very similar to #189 with the seed net etc, just more samples and the techniques above, and we will see if we are able to beat next :)

### CODE REVIEW

**Comment by jorgenhanssen** [comment]
984 million samples generated. Gonna try a 0.4 vs 0.5 `draw target` for 30 epochs each on the same seed and net. then i will run a full run of whichever is best.

**Comment by jorgenhanssen** [comment]
Not the jump I was hoping for, but ok.

**Comment by jorgenhanssen** [comment]
## Iteration 9

`./target/release/generate random --plies 8 --nodes 15000 --max-opening-imbalance 600 --syzygy-path "$HOME/syzygy/3-4-5:$HOME/syzygy/6-WDL"`
`./target/release/train --wdl 0.2 --restart`

```
Files:                 31
Threshold:             600 cp
Total samples:         987,754,560

Draw disagreements: 9,560,501 (0.968%)
   -799  D  4k3/R3P1p1/2p1P3/P7/8/7p/4K3/7r w - - 0 49
   +631  D  1b1rk2r/pp4p1/2q1p3/2p1n1Np/3n4/P1N4P/1PPB2P1/3RQR1K w k - 6 19
   -627  D  6r1/4nQ2/3kPRp1/2p4p/1p1p4/3P3P/6PK/4qN2 b - - 1 45
   -647  D  1Q2r3/N1k5/5pp1/3R1n2/R4K1p/p1P4P/Pr6/8 b - - 0 50
   -618  D  2R5/7P/5N2/4n3/pp3p2/k4P2/4K1Pr/8 b - - 7 51

Decisive opposites: 295,199 (0.030%)
   +724  B  2r2k2/3bn1bQ/3p2P1/1p2p3/2qP1p2/P1P1BB2/P5P1/1N2R1K1 w - - 0 36
   -613  W  1k1r3r/1p3p2/p3bqp1/2p1p1pP/1nPpP1P1/1P1P2NR/3QBPK1/7R b - - 0 33
   -684  W  r4k1r/1q1n4/3b1pnP/2pbp3/p7/P1BP1N1R/1P3PB1/1KQ1R3 w - - 0 29
   +726  B  2k3r1/1pp5/4np1P/p6K/P1Rb2P1/1r1N1R2/1P6/2B5 w - - 5 48
   +620  B  r3k1r1/1p1q1pb1/p1np3p/3Ppb2/3PB3/1P6/PBN1QP2/1K1R3R b - - 0 21
```

```
06:53:46 [INFO] Total samples: 987754560
06:53:46 [INFO] Unique positions: ~100.24%
06:53:46 [INFO] Total games: 6461363
06:53:46 [INFO] Outcomes: white 27.8% / draw 44.8% / black 27.4%
06:53:46 [INFO] Output bucket distribution:
06:53:46 [INFO]   Bucket 0: 113011503 samples (11.4%)
06:53:46 [INFO]   Bucket 1: 200633357 samples (20.3%)
06:53:46 [INFO]   Bucket 2: 159327592 samples (16.1%)
06:53:46 [INFO]   Bucket 3: 116369570 samples (11.8%)
06:53:46 [INFO]   Bucket 4: 98288960 samples (10.0%)
06:53:46 [INFO]   Bucket 5: 94412143 samples (9.6%)
06:53:46 [INFO]   Bucket 6: 101458807 samples (10.3%)
06:53:46 [INFO]   Bucket 7: 104252628 samples (10.6%)
```

```
Depth        : 14
Max seldepth : 44
Avg seldepth : 27.3
Total time   : 9904 ms
Total nodes  : 14212096
Avg NPS      : 1434985
Avg branching: 2.40
Bench hash   : 03F0271C0A45B925
```

```
--------------------------------------------------
Results of grail vs grail-next (60000 nodes, 1t, 256MB, UHO_Lichess_4852_v1.epd):
Elo: -23.31 +/- 14.65, nElo: -34.40 +/- 21.53
LOS: 0.09 %, DrawRatio: 46.20 %, PairsRatio: 0.78
Games: 1000, Wins: 317, Losses: 384, Draws: 299, Points: 466.5 (46.65 %)
Ptnml(0-2): [50, 101, 231, 102, 16], WL/DD Ratio: 3.81
--------------------------------------------------
```

**Comment by jorgenhanssen** [comment]
Training iteration 10 now on 1.25B samples

**Comment by jorgenhanssen** [comment]
## The Revelation

Currently training iteration 10 on epoch 73. It is still losing convincingly to `next`:

```
--------------------------------------------------
Results of grail vs grail-next (60000 nodes, 1t, 16MB, UHO_Lichess_4852_v1.epd):
Elo: -30.65 +/- 16.00, nElo: -41.51 +/- 21.53
LOS: 0.01 %, DrawRatio: 40.00 %, PairsRatio: 0.64
Games: 1000, Wins: 307, Losses: 395, Draws: 298, Points: 456.0 (45.60 %)
Ptnml(0-2): [54, 129, 200, 85, 32], WL/DD Ratio: 3.76
--------------------------------------------------
```

Then it hit me.. `next` is trained on data generated solely from UHO openings, so duuh of course it will have an edge when the same book is used for SPRT...

image

I generated 1M random 8-ply openings as an EPD and sure enough:

```
--------------------------------------------------
Results of grail vs grail-next (60000 nodes, 1t, 16MB, random-8.epd):
Elo: 6.95 +/- 12.84, nElo: 11.66 +/- 21.53
LOS: 85.57 %, DrawRatio: 61.20 %, PairsRatio: 1.18
Games: 1000, Wins: 405, Losses: 385, Draws: 210, Points: 510.0 (51.00 %)
Ptnml(0-2): [25, 64, 306, 76, 29], WL/DD Ratio: 7.74
--------------------------------------------------
```

## So what now?

Well. imo, `next` is lowkey cheating. Or, at least it is specialized towards UHO books in a way I don't like, because these are the books that are often used in CCRL etc, and to me this feels like a shitty advantage.

But! This net is not _that_ far behind, only 30 Elo. And on truly random openings it appears to be slightly ahead.

My plan forward is to let iteration 10 finish. Then I will likely merge it if it holds up reasonably in an SPRT on random opening data. That being said, I don't want the next version of grail to regress against 2.0.1 though, so I might end up "fine-tuning" the net on UHO books or FRC positions (if I end up adding that to v3) later but as a separate PR.

**Comment by jorgenhanssen** [comment]
## Post-mortem

```
--------------------------------------------------
Results of grail vs grail-next (60000 nodes, 1t, 256MB, UHO_Lichess_4852_v1.epd):
Elo: -41.68 +/- 6.94, nElo: -58.41 +/- 9.63
LOS: 0.00 %, DrawRatio: 41.64 %, PairsRatio: 0.56
Games: 5000, Wins: 1446, Losses: 2043, Draws: 1511, Points: 2201.5 (44.03 %)
Ptnml(0-2): [293, 643, 1041, 414, 109], WL/DD Ratio: 3.59
--------------------------------------------------
```

```
--------------------------------------------------
Results of grail vs grail-next (60000 nodes, 1t, 256MB, random-8.epd):
Elo: -8.55 +/- 6.03, nElo: -13.65 +/- 9.63
LOS: 0.27 %, DrawRatio: 59.08 %, PairsRatio: 0.85
Games: 5000, Wins: 1939, Losses: 2062, Draws: 999, Points: 2438.5 (48.77 %)
Ptnml(0-2): [176, 378, 1477, 331, 138], WL/DD Ratio: 9.19
--------------------------------------------------
```

Closing this branch. Will experiment more with datagen using the current net instead.
